Gradle build fails if specifying productionMode=true when using a custom sourceset

### Description of the bug

When running a Gradle build with `productionMode = true`, if the Vaadin dependencies are not specified against the `main`  sourceset (e.g. an additional sourceset named `ui` is created and the dependencies listed against `uiImplementation`), or the classes that use Vaadin components are not in the `main` sourceset then the build fails with Vite reporting that dependencies could not be found, e.g.

```
Execution failed for task ':vaadinBuildFrontend'.
> Vite process exited with non-zero exit code.
  Stderr: '[36mvite v4.0.2 [32mbuilding for production...[36m[39m
  transforming...
  Use of eval in "frontend/generated/jar-resources/FlowClient.js" is strongly discouraged as it poses security risks and may cause issues with minification.
  Use of eval in "frontend/generated/jar-resources/FlowClient.js" is strongly discouraged as it poses security risks and may cause issues with minification.
  [32m?[39m 37 modules transformed.
  rendering chunks...
  computing gzip size...
  [2m../build/resources/main/META-INF/VAADIN/webapp/[22m[32mindex.html                                                [39m[1m[2m  0.60 kB[22m[1m[22m
  [2m../build/resources/main/META-INF/VAADIN/webapp/[22m[2mVAADIN/build/[22m[36mgenerated-flow-imports-fallback-e5b0ae68.js  [39m[1m[2m  0.19 kB[22m[1m[22m[2m ? gzip:  0.17 kB[22m
  [2m../build/resources/main/META-INF/VAADIN/webapp/[22m[2mVAADIN/build/[22m[36mgenerated-flow-imports-fc2941af.js           [39m[1m[2m  0.83 kB[22m[1m[22m[2m ? gzip:  0.49 kB[22m
  [2m../build/resources/main/META-INF/VAADIN/webapp/[22m[2mVAADIN/build/[22m[36mFlowBootstrap-feff2646.js                    [39m[1m[2m  3.42 kB[22m[1m[22m[2m ? gzip:  1.45 kB[22m
  [2m../build/resources/main/META-INF/VAADIN/webapp/[22m[2mVAADIN/build/[22m[36mindexhtml-fb749209.js                        [39m[1m[2m 59.60 kB[22m[1m[22m[2m ? gzip: 19.52 kB[22m
  [2m../build/resources/main/META-INF/VAADIN/webapp/[22m[2mVAADIN/build/[22m[36mFlowClient-984d8da1.js                       [39m[1m[2m147.31 kB[22m[1m[22m[2m ? gzip: 49.31 kB[22m
  frontend/generated/jar-resources/lumo-includes.ts(1,23): error TS2307: Cannot find module '@vaadin/vaadin-lumo-styles/color.js' or its corresponding type declarations.
  frontend/generated/jar-resources/lumo-includes.ts(2,28): error TS2307: Cannot find module '@vaadin/vaadin-lumo-styles/typography.js' or its corresponding type declarations.
  frontend/generated/jar-resources/material-includes.ts(1,28): error TS2307: Cannot find module '@vaadin/vaadin-material-styles' or its corresponding type declarations.
  frontend/generated/jar-resources/material-includes.ts(2,28): error TS2307: Cannot find module '@vaadin/vaadin-material-styles' or its corresponding type declarations.
  frontend/generated/jar-resources/tooltip.ts(1,25): error TS2307: Cannot find module '@vaadin/tooltip' or its corresponding type declarations.
  '
```

### Expected behavior

The `vaadinPrepareFrontend` task should successfully copy supporting Javascript assets detected for components in the non-main sourceset into the `frontend/generated/jar-resources/` directory, and the `vaadinBuildFrontend` task should successfully complete, producing a valid JAR that contains a working Vaadin distribution.

### Minimal reproducible example

Create a Gradle project with the following contents:

* build.gradle
```gradle
plugins {
    id 'java'
    id 'org.springframework.boot' version '3.0.0'
    id 'io.spring.dependency-management' version '1.0.11.RELEASE'
    id 'com.vaadin' version '23.3.4'
}

repositories {
    mavenLocal()
    mavenCentral()
    maven { url 'https://maven.vaadin.com/vaadin-prereleases' }
    maven { url 'https://repo.spring.io/milestone' }
}

sourceSets {
    ui {
        java
    }
    main {
        java {
            compileClasspath += ui.output
            runtimeClasspath += ui.output + ui.runtimeClasspath
        }
    }
}

vaadin {
    productionMode = true
}

dependencies {
    uiImplementation('com.vaadin:vaadin-spring-boot-starter')
    implementation('org.springframework.boot:spring-boot-starter-web')
}

dependencyManagement {
    uiImplementation {
        imports {
            mavenBom "com.vaadin:vaadin-bom:23.3.4"
        }
    }
}

jar {
    enabled = false
}
```

* src/main/java/com/example/demo/DemoApplication.java
```java
package com.example.demo;
            
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class DemoApplication {

    public static void main(String[] args) {
        SpringApplication.run(DemoApplication.class, args);
    }

}
```

* src/ui/java/com/example/demo/AppShell.java
```java
package com.example.demo;
            
import com.vaadin.flow.component.page.AppShellConfigurator;
import com.vaadin.flow.server.PWA;

@PWA(name = "Demo application", shortName = "Demo")
public class AppShell implements AppShellConfigurator {
}
```

Run `gradle build`

### Versions

- Vaadin / Flow version: 23.3.4
- Java version: 17.0.4.1 Corretto
- OS version: Windows 10 (although not OS specific)
- Browser version (if applicable): n/a
- Application Server (if applicable): n/a
- IDE (if applicable): n/a

--- a/flow-plugins/flow-gradle-plugin/src/functionalTest/kotlin/com/vaadin/gradle/MiscSingleModuleTest.kt
+++ b/flow-plugins/flow-gradle-plugin/src/functionalTest/kotlin/com/vaadin/gradle/MiscSingleModuleTest.kt
@@ -1,5 +1,5 @@
 /**
- *    Copyright 2000-2022 Vaadin Ltd
+ *    Copyright 2000-2023 Vaadin Ltd
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -531,5 +531,98 @@ class MiscSingleModuleTest : AbstractGradleTest() {
         expect(listOf("flow-client-", "flow-commons-upload-", "flow-html-components-", "flow-lit-template-", "flow-push-", "flow-server-")) {
             classpath.map { it.removeSuffix("-SNAPSHOT.jar").dropLastWhile { it != '-' } }
         }
+    }
+
+    @Test
+    fun testUsingNonMainSourceSet() {
+        testProject.settingsFile.writeText(
+            """
+            pluginManagement {
+                repositories {
+                  gradlePluginPortal()
+                }
+              }
+            """
+        )
+        testProject.buildFile.writeText(
+            """
+            plugins {
+                id 'java'
+                id 'org.springframework.boot' version '3.0.0'
+                id 'io.spring.dependency-management' version '1.0.11.RELEASE'
+                id("com.vaadin")
+            }
+            
+            repositories {
+                mavenLocal()
+                mavenCentral()
+                maven { url 'https://maven.vaadin.com/vaadin-prereleases' }
+            }
+            
+            sourceSets {
+                ui {
+                    java
+                }
+                main {
+                    java {
+                        compileClasspath += ui.output
+                        runtimeClasspath += ui.output + ui.runtimeClasspath
+                    }
+                }
+            }
+            
+            vaadin {
+                productionMode = true
+                sourceSetName = 'ui'
+            }
+            
+            dependencies {
+                uiImplementation('com.vaadin:flow:$flowVersion')
+                implementation('org.springframework.boot:spring-boot-starter-web')
+            }
+            
+            jar {
+                enabled = false
+            }
+            """
+        )
+
+        testProject.newFile(
+            "src/main/java/com/example/demo/DemoApplication.java", """
+            package com.example.demo;
+            
+            import org.springframework.boot.SpringApplication;
+            import org.springframework.boot.autoconfigure.SpringBootApplication;
+            
+            @SpringBootApplication
+            public class DemoApplication {
+            
+                public static void main(String[] args) {
+                    SpringApplication.run(DemoApplication.class, args);
+                }
+            
+            }
+        """.trimIndent()
+        )
+
+        testProject.newFile(
+            "src/ui/java/com/example/demo/AppShell.java", """
+            package com.example.demo;
+            
+            import com.vaadin.flow.component.page.AppShellConfigurator;
+            import com.vaadin.flow.server.PWA;
+
+            @PWA(name = "Demo application", shortName = "Demo")
+            public class AppShell implements AppShellConfigurator {
+            }
+        """.trimIndent()
+        )
+
+        val build: BuildResult = testProject.build("build")
+        build.expectTaskSucceded("vaadinPrepareFrontend")
+        build.expectTaskSucceded("vaadinBuildFrontend")
+
+        val jar: File = testProject.builtJar
+        expectArchiveContainsVaadinBundle(jar, true)
     }
 }

--- a/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/GradlePluginAdapter.kt
+++ b/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/GradlePluginAdapter.kt
@@ -1,5 +1,5 @@
 /*
- * Copyright 2000-2022 Vaadin Ltd
+ * Copyright 2000-2023 Vaadin Ltd
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -45,10 +45,10 @@ internal class GradlePluginAdapter(val project: Project, private val isBeforePro
     override fun generatedTsFolder(): File = extension.generatedTsFolder
 
     override fun getClassFinder(): ClassFinder {
-        val runtimeClasspath: Configuration? = project.configurations.findByName("runtimeClasspath")
-        val runtimeClasspathJars: List<File> = if (runtimeClasspath != null) {
+        val dependencyConfiguration: Configuration? = project.configurations.findByName(extension.dependencyScope!!)
+        val dependencyConfigurationJars: List<File> = if (dependencyConfiguration != null) {
             var artifacts: List<ResolvedArtifact> =
-                runtimeClasspath.resolvedConfiguration.resolvedArtifacts.toList()
+                dependencyConfiguration.resolvedConfiguration.resolvedArtifacts.toList()
             val extension = VaadinFlowPluginExtension.get(project)
             val artifactFilter = extension.classpathFilter.toPredicate()
             artifacts = artifacts.filter { artifactFilter.test(it.moduleVersion.id.module) }
@@ -57,11 +57,11 @@ internal class GradlePluginAdapter(val project: Project, private val isBeforePro
 
         // we need to also analyze the project's classes
         val sourceSet: SourceSetContainer = project.properties["sourceSets"] as SourceSetContainer
-        val classesDirs: List<File> = sourceSet.getByName("main").output.classesDirs
+        val classesDirs: List<File> = sourceSet.getByName(extension.sourceSetName).output.classesDirs
             .toList()
             .filter { it.exists() }
 
-        val resourcesDir: List<File> = listOfNotNull(sourceSet.getByName("main").output.resourcesDir)
+        val resourcesDir: List<File> = listOfNotNull(sourceSet.getByName(extension.sourceSetName).output.resourcesDir)
                 .filter { it.exists() }
 
         // for Spring Boot project there is no "providedCompile" scope: the WAR plugin brings that in.
@@ -71,7 +71,7 @@ internal class GradlePluginAdapter(val project: Project, private val isBeforePro
             ?.toList()
             ?: listOf()
 
-        val apis: Set<File> = (runtimeClasspathJars + classesDirs + resourcesDir + servletJar).toSet()
+        val apis: Set<File> = (dependencyConfigurationJars + classesDirs + resourcesDir + servletJar).toSet()
 
         // eagerly check that all the files/folders exist, to avoid spamming the console later on
         // see https://github.com/vaadin/vaadin-gradle-plugin/issues/38 for more details
@@ -94,7 +94,7 @@ internal class GradlePluginAdapter(val project: Project, private val isBeforePro
     }
 
     override fun getJarFiles(): MutableSet<File> {
-        val jarFiles: Set<File> = project.configurations.runtimeClasspath.jars.toSet()
+        val jarFiles: Set<File> = project.configurations.getByName(extension.dependencyScope!!).jars.toSet()
         return jarFiles.toMutableSet()
     }
 

--- a/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinFlowPluginExtension.kt
+++ b/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinFlowPluginExtension.kt
@@ -1,5 +1,5 @@
 /**
- *    Copyright 2000-2022 Vaadin Ltd
+ *    Copyright 2000-2023 Vaadin Ltd
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -18,7 +18,6 @@ package com.vaadin.gradle
 import com.vaadin.flow.server.Constants
 import com.vaadin.flow.server.InitParameters
 import com.vaadin.flow.server.frontend.FrontendTools
-import com.vaadin.flow.server.frontend.FrontendUtils
 import com.vaadin.flow.server.frontend.installer.NodeInstaller
 import groovy.lang.Closure
 import groovy.lang.DelegatesTo
@@ -190,6 +189,27 @@ public open class VaadinFlowPluginExtension(project: Project) {
 
     public var classpathFilter: ClasspathFilter = ClasspathFilter()
 
+    /**
+     * The name of the SourceSet to scan for Vaadin components - i.e. the classes that are annoated with
+     * Vaadin annotations.
+     */
+    public var sourceSetName : String = "main"
+
+    /**
+     * The Gradle scope the Vaadin dependencies have been added to. Defaults to 'runtimeClasspath' if
+     * no sourceSetName has been specified, or '<code>sourceSetName</code>RuntimeClasspath' if a non-main sourceset
+     * has been set.
+     */
+    public var dependencyScope : String? = null
+
+    /**
+     * The Gradle task that the `vaadinPrepareFrontend` task must run before. The target task should run before
+     * or be the task that copies the files from the resources directories of the specified SourceSet to the relevant
+     * output directory for that SourceSet. Defaults to 'processResources' if no sourceSetName has been specified, or
+     * 'process<code>SourceSetName</code>Resources' if a non-main sourceset has been specified.
+     */
+    public var processResourcesTaskName : String? = null
+
     public fun filterClasspath(@DelegatesTo(value = ClasspathFilter::class, strategy = Closure.DELEGATE_FIRST) block: Closure<*>? = null): ClasspathFilter {
         if (block != null) {
             block.delegate = classpathFilter
@@ -234,6 +254,24 @@ public open class VaadinFlowPluginExtension(project: Project) {
         if (useGlobalPnpmProperty != null) {
             useGlobalPnpm = useGlobalPnpmProperty
         }
+
+        // calculate processResourcesTaskName if not set by user
+        if (processResourcesTaskName == null) {
+            processResourcesTaskName = if (sourceSetName == "main") {
+                "processResources"
+            } else {
+                "process${sourceSetName.replaceFirstChar(Char::titlecase)}Resources"
+            }
+        }
+
+        // calculate dependencyScope if not set by user
+        if (dependencyScope == null) {
+            dependencyScope = if (sourceSetName == "main") {
+                "runtimeClasspath"
+            } else  {
+                sourceSetName + "RuntimeClasspath"
+            }
+        }
     }
 
     override fun toString(): String = "VaadinFlowPluginExtension(" +
@@ -258,13 +296,16 @@ public open class VaadinFlowPluginExtension(project: Project) {
             "generatedTsFolder=$generatedTsFolder, " +
             "nodeVersion=$nodeVersion, " +
             "nodeDownloadRoot=$nodeDownloadRoot, " +
-            "nodeAutoUpdate=$nodeAutoUpdate" +
-            "resourceOutputDirectory=$resourceOutputDirectory" +
-            "postinstallPackages=$postinstallPackages" +
+            "nodeAutoUpdate=$nodeAutoUpdate, " +
+            "resourceOutputDirectory=$resourceOutputDirectory, " +
+            "postinstallPackages=$postinstallPackages, " +
+            "sourceSetName=$sourceSetName, " +
+            "dependencyScope=$dependencyScope, " +
+            "processResourcesTaskName=$processResourcesTaskName" +
             ")"
 }
 
 internal val Project.buildResourcesDir: File get() {
     val sourceSets: SourceSetContainer = project.convention.getPlugin(JavaPluginConvention::class.java).sourceSets
-    return sourceSets.getByName("main").output.resourcesDir!!
+    return sourceSets.getByName(extensions.getByType(VaadinFlowPluginExtension::class.java).sourceSetName).output.resourcesDir!!
 }

--- a/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinPlugin.kt
+++ b/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinPlugin.kt
@@ -1,5 +1,5 @@
 /**
- *    Copyright 2000-2022 Vaadin Ltd
+ *    Copyright 2000-2023 Vaadin Ltd
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -49,10 +49,10 @@ public class VaadinPlugin : Plugin<Project> {
 
             // add a new source-set folder for generated stuff, by default `vaadin-generated`
             val sourceSets: SourceSetContainer = it.properties["sourceSets"] as SourceSetContainer
-            sourceSets.getByName("main").resources.srcDirs(extension.resourceOutputDirectory)
+            sourceSets.getByName(extension.sourceSetName).resources.srcDirs(extension.resourceOutputDirectory)
 
             // auto-activate tasks: https://github.com/vaadin/vaadin-gradle-plugin/issues/48
-            project.tasks.getByPath("processResources").dependsOn("vaadinPrepareFrontend")
+            project.tasks.getByPath(extension.processResourcesTaskName!!).dependsOn("vaadinPrepareFrontend")
             if (extension.productionMode) {
                 // this will also catch the War task since it extends from Jar
                 project.tasks.withType(Jar::class.java) { task: Jar ->

--- a/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinPrepareFrontendTask.kt
+++ b/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinPrepareFrontendTask.kt
@@ -1,5 +1,5 @@
 /**
- *    Copyright 2000-2022 Vaadin Ltd
+ *    Copyright 2000-2023 Vaadin Ltd
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -18,8 +18,6 @@ package com.vaadin.gradle
 import com.vaadin.flow.plugin.base.BuildFrontendUtil
 import com.vaadin.flow.server.frontend.FrontendTools
 import org.gradle.api.DefaultTask
-import org.gradle.api.Task
-import org.gradle.api.artifacts.ProjectDependency
 import org.gradle.api.tasks.TaskAction
 
 /**
@@ -33,17 +31,18 @@ public open class VaadinPrepareFrontendTask : DefaultTask() {
         group = "Vaadin"
         description = "checks that node and npm tools are installed, copies frontend resources available inside `.jar` dependencies to `node_modules`, and creates or updates `package.json` and `webpack.config.json` files."
 
+        val extension: VaadinFlowPluginExtension = VaadinFlowPluginExtension.get(project)
         // Maven's task run in the LifecyclePhase.PROCESS_RESOURCES phase
 
         // the processResources copies stuff from build/vaadin-generated
         // (which is populated by this task) and therefore must run after this task.
-        project.tasks.getByName("processResources").mustRunAfter("vaadinPrepareFrontend")
+        project.tasks.getByName(extension.processResourcesTaskName!!).mustRunAfter("vaadinPrepareFrontend")
 
         // make sure all dependent projects have finished building their jars, otherwise
         // the Vaadin classpath scanning will not work properly. See
         // https://github.com/vaadin/vaadin-gradle-plugin/issues/38
         // for more details.
-        dependsOn(project.configurations.runtimeClasspath.jars)
+        dependsOn(project.configurations.getByName(extension.dependencyScope!!).jars)
     }
 
     @TaskAction

--- a/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinUtils.kt
+++ b/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinUtils.kt
@@ -1,5 +1,5 @@
 /**
- *    Copyright 2000-2022 Vaadin Ltd
+ *    Copyright 2000-2023 Vaadin Ltd
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -17,7 +17,6 @@ package com.vaadin.gradle
 
 
 import org.gradle.api.artifacts.Configuration
-import org.gradle.api.artifacts.ConfigurationContainer
 import org.gradle.api.file.FileCollection
 import com.vaadin.flow.function.SerializableSupplier
 import com.vaadin.flow.server.frontend.FrontendTools
@@ -73,12 +72,6 @@ internal fun VaadinFlowPluginExtension.createFrontendTools(): FrontendTools {
     settings.setNodeDownloadRoot(URI(nodeDownloadRoot))
     return FrontendTools(settings)
 }
-
-/**
- * Returns the "runtimeClasspath" file collection.
- */
-internal val ConfigurationContainer.runtimeClasspath: Configuration
-    get() = getByName("runtimeClasspath")
 
 /**
  * Returns only jar files from given file collection.

